src/unlimited.cpp

```diff
@@ -1615,6 +1618,222 @@ UniValue setlog(const UniValue &params, bool fHelp)
+    }
+    mkblocktemplate(params, &candid.block);
+    ret = MkMiningCandidateJson(candid);
+    return ret;
+}
+
+/** RPC Submit a solved block candidate*/
+UniValue submitminingsolution(const UniValue &params, bool fHelp)
+{
+    UniValue rcvd;
+    CBlock block;
+    LOCK(cs_main);
+
+    if (fHelp || params.size() != 1)
+    {
+        throw runtime_error(
+            "submitminingsolution \"Mining-Candidate data\" ( \"jsonparametersobject\" )\n"
+            "\nAttempts to submit a new block to the network.\n"
+            "\nArguments\n"
+            "1. \"submitminingsolutiondata\"    (string, required) the mining solution (JSON encoded) data to submit\n"
+            "\nResult:\n"
+            "\nNothing on success, error string if block was rejected.\n"
+            "Identical to \"submitblock\".\n"
+            "\nExamples:\n" +
+            HelpExampleRpc("submitminingsolution", "\"mydata\""));
+    }
+
+    rcvd = params[0].get_obj();
+
+    int64_t id = rcvd["id"].get_int64();
+
+    // Needs LOCK(cs_main); above:
+    if (miningCandidatesMap.count(id) == 1)
+    {
+        block = miningCandidatesMap[id].block;
+        miningCandidatesMap.erase(id);
+    }
+    else
+    {
+        return UniValue("id not found");
+    }
+
+    UniValue nonce = rcvd["nonce"];
+    if (nonce.isNull())
+    {
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "nonce not found");
+    }
+    block.nNonce = (uint32_t)nonce.get_int64(); // 64 bit to deal with sign bit in 32 bit unsigned int
+
+    UniValue time = rcvd["time"];
+    if (!time.isNull())
+    {
+        block.nTime = (uint32_t)time.get_int64();
+    }
+
+    UniValue version = rcvd["version"];
+    if (!version.isNull())
+    {
+        block.nVersion = version.get_int(); // version signed 32 bit int
+    }
+
+    // Coinbase:
+    CTransaction coinbase;
+    UniValue cbhex = rcvd["coinbase"];
+    if (!cbhex.isNull())
+    {
+        if (DecodeHexTx(coinbase, cbhex.get_str()))
+            block.vtx[0] = MakeTransactionRef(std::move(coinbase));
+        else
+        {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "coinbase decode failed");
+        }
+    }
+
+    // MerkleRoot:
+    {
+        std::vector<uint256> merklebranches = GetMerkleProofBranches(&block);
+        uint256 t = block.vtx[0]->GetHash();
+        block.hashMerkleRoot = CalculateMerkleRoot(t, merklebranches);
+    }
+
+    UniValue uvsub = SubmitBlock(block); // returns string on failure
+    RmOldMiningCandidates();
+    return uvsub;
+}
+
+static void CalculateNextMerkleRoot(uint256 &merkle_root, uint256 &merkle_branch)
+{
+    // Append a branch to the root. Double SHA256 the whole thing:
+    uint256 hash;
+    CHash256()
+        .Write(merkle_root.begin(), merkle_root.size())
+        .Write(merkle_branch.begin(), merkle_branch.size())
+        .Finalize(hash.begin());
+    merkle_root = hash;
+}
+
+uint256 CalculateMerkleRoot(uint256 &coinbase_hash, std::vector<uint256> merklebranches)
+{
```

Comment:
pass reference  "std::vector<uint256> &merklebranches"

Reply:
Thanks I missed that.